client/hotrod-client/src/test/java/org/infinispan/client/hotrod/counter/BaseCounterAPITest.java

```diff
@@ -0,0 +1,204 @@
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+
+import org.infinispan.client.hotrod.counter.impl.ConnectionManager;
+import org.infinispan.client.hotrod.counter.impl.NotificationManager;
+import org.infinispan.client.hotrod.counter.impl.RemoteCounterManager;
+import org.infinispan.commons.util.ReflectionUtil;
+import org.infinispan.counter.api.CounterEvent;
+import org.infinispan.counter.api.CounterListener;
+import org.infinispan.counter.api.Handle;
+import org.infinispan.server.hotrod.counter.impl.BaseCounterImplTest;
+
+/**
+ * A base test class for {@link org.infinispan.counter.api.StrongCounter} and {@link
+ * org.infinispan.counter.api.WeakCounter} implementation in the Hot Rod client.
+ *
+ * @author Pedro Ruivo
+ * @since 9.2
+ */
+public abstract class BaseCounterAPITest<T> extends AbstractCounterTest {
+
+   private static final CounterListener EXCEPTION_LISTENER = entry -> {
+      throw new RuntimeException("induced");
+   };
+
+   public void testExceptionInListener(Method method) throws InterruptedException {
+      final String counterName = method.getName();
+      T counter = defineAndCreateCounter(counterName, 0);
+
+      Handle<BaseCounterImplTest.EventLogger> handle1 = addListenerTo(counter, new BaseCounterImplTest.EventLogger());
+      Handle<CounterListener> handleEx = addListenerTo(counter, EXCEPTION_LISTENER);
+      Handle<BaseCounterImplTest.EventLogger> handle2 = addListenerTo(counter, new BaseCounterImplTest.EventLogger());
+
+      add(counter, 1, 1);
+      add(counter, -1, 0);
+      add(counter, 10, 10);
+      add(counter, 1, 11);
+      add(counter, 2, 13);
+
+      assertNextValidEvent(handle1, 0, 1);
+      assertNextValidEvent(handle1, 1, 0);
+      assertNextValidEvent(handle1, 0, 10);
+      assertNextValidEvent(handle1, 10, 11);
+      assertNextValidEvent(handle1, 11, 13);
+
+      assertNextValidEvent(handle2, 0, 1);
+      assertNextValidEvent(handle2, 1, 0);
+      assertNextValidEvent(handle2, 0, 10);
+      assertNextValidEvent(handle2, 10, 11);
+      assertNextValidEvent(handle2, 11, 13);
+
+
+      assertNoEvents(handle1);
+      assertNoEvents(handle2);
+
+      handle1.remove();
+      handle2.remove();
+      handleEx.remove();
+   }
+
+   public void testConcurrentListenerAddAndRemove(Method method) throws InterruptedException, ExecutionException {
+      String counterName = method.getName();
+      defineAndCreateCounter(counterName, 1);
+      List<T> counters = getCounters(counterName);
+
+      List<IncrementTask> taskList = counters.stream()
+            .map(IncrementTask::new)
+            .collect(Collectors.toList());
+
+      List<Future<?>> futureTaskList = taskList.stream()
+            .map(this::fork)
+            .collect(Collectors.toList());
+
+      T counter = counters.get(0);
+
+      Handle<BaseCounterImplTest.EventLogger> handle = addListenerTo(counter, new BaseCounterImplTest.EventLogger());
+      Thread.sleep(1000);
```

Comment:
:-( ?

Reply:
just to give some time to collect some events